File: Components/WinForms/VB/GridTutorials/GridDragDropInternal/GridDragDropInternal.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Drawing
Imports System.Collections
Imports System.ComponentModel
Imports System.Windows.Forms
Imports System.Data
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.XtraGrid.Demos.Tutorials
    ''' <summary>
    ''' Summary description for GridDragDropInternal.
    ''' </summary>
    Partial Public Class GridDragDropInternal
        Inherits TutorialControl
        Public Sub New()
            '
            ' Required for Windows Form Designer support
            '
            InitializeComponent()

            '
            ' TODO: Add any constructor code after InitializeComponent call
            '
        End Sub
        Private hitInfo As GridHitInfo = Nothing
        Private Sub GridDragDropInternal_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles MyBase.Load
            InitData()
            gridView1.SetRowExpanded(-1, True)
        End Sub

        Private Sub InitData()
            Dim DBFileName As String = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath, "Data\Contacts.xml")
            If DBFileName <> "" Then
                Dim ds As DataSet = New DataSet()
                ds.ReadXml(DBFileName)
                gridControl1.DataSource = ds.Tables(0)
            End If
        End Sub


        '<gridControl1>
        Private Sub gridControl1_MouseDown(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles gridControl1.MouseDown
            hitInfo = gridView1.CalcHitInfo(New Point(e.X, e.Y))
            If hitInfo.RowHandle < 0 Then
                hitInfo = Nothing
            End If
        End Sub

        Private Sub gridControl1_MouseMove(ByVal sender As Object, ByVal e As System.Windows.Forms.MouseEventArgs) Handles gridControl1.MouseMove
            If hitInfo Is Nothing Then
                Return
            End If
            If e.Button <> Windows.Forms.MouseButtons.Left Then
                Return
            End If
            Dim dragRect As Rectangle = New Rectangle(New Point(hitInfo.HitPoint.X - SystemInformation.DragSize.Width \ 2, hitInfo.HitPoint.Y - SystemInformation.DragSize.Height \ 2), SystemInformation.DragSize)
            If (Not dragRect.Contains(New Point(e.X, e.Y))) Then
                If hitInfo.InRow Then
                    gridControl1.DoDragDrop(New Object() {hitInfo.RowHandle, gridControl1}, DragDropEffects.All)
                End If
            End If
        End Sub
        '</gridControl1>

        Private Function GetDragObject(ByVal sender As Object, ByVal data As IDataObject) As Integer
            Dim obj As Object() = TryCast(data.GetData(GetType(Object())), Object())
            If obj Is Nothing OrElse (Not sender.Equals(gridControl1)) Then
                Return -2
            End If
            If TypeOf obj(0) Is Int32 Then
                Return CInt(Fix(obj(0)))
            Else
                Return -1
            End If
        End Function

        '<gridControl1>
        Private Sub gridControl1_DragOver(ByVal sender As Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles gridControl1.DragOver
            Dim hi As GridHitInfo = gridView1.CalcHitInfo(gridControl1.PointToClient(New Point(e.X, e.Y)))
            Dim handle As Integer = GetDragObject(sender, e.Data)
            If hi.InRow AndAlso handle >= 0 AndAlso handle <> hi.RowHandle Then
                If (e.KeyState And 8) <> 0 Then
                    e.Effect = DragDropEffects.Copy
                Else
                    e.Effect = DragDropEffects.Move
                End If
            Else
                e.Effect = DragDropEffects.None
            End If
        End Sub

        Private Sub gridControl1_DragDrop(ByVal sender As Object, ByVal e As System.Windows.Forms.DragEventArgs) Handles gridControl1.DragDrop
            Dim hi As GridHitInfo = gridView1.CalcHitInfo(gridControl1.PointToClient(New Point(e.X, e.Y)))
            Dim handle As Integer = GetDragObject(sender, e.Data)
            If handle < 0 Then
                Return
            End If
            If hi.InRow Then
                Dim i As Integer = 0
                Dim row As DataRowView = TryCast(gridView1.GetRow(handle), DataRowView)
                'the following code determines whether the target row is a group row and if a group row is detected, 
                'it retrieves the actual data row corresponding to it.
                Dim newHandle As Integer = hi.RowHandle
                If newHandle < 0 Then
                    gridView1.SetRowExpanded(newHandle, True, True)
                    newHandle = gridView1.GetDataRowHandleByGroupRowHandle(newHandle)
                End If
                Dim rowReciever As DataRow = gridView1.GetDataRow(newHandle)
                'the following code implements actual data moving or copying
                Dim newRow As DataRowView = Nothing
                If e.Effect = DragDropEffects.Copy Then
                    newRow = (CType(gridView1.DataSource, DataView)).AddNew()
                    i = 0
                    Do While i < gridView1.Columns.Count
                        newRow.Row(gridView1.Columns(i).FieldName) = row(gridView1.Columns(i).FieldName)
                        i += 1
                    Loop
                    row = newRow
                End If
                i = 0
                Do While i < gridView1.SortInfo.GroupCount
                    row.Row(gridView1.GroupedColumns(i).FieldName) = rowReciever(gridView1.GroupedColumns(i).FieldName)
                    i += 1
                Loop
                If Not newRow Is Nothing Then
                    newRow.Row("id") = (CType(gridView1.DataSource, DataView)).Count
                End If
                row.EndEdit()
            End If
        End Sub
        '</gridControl1>

        Private Sub gridView1_ShowingEditor(ByVal sender As Object, ByVal e As System.ComponentModel.CancelEventArgs) Handles gridView1.ShowingEditor
            If (Not MouseButtons.Equals(Windows.Forms.MouseButtons.None)) Then
                e.Cancel = True
            End If
        End Sub
        Public Overrides ReadOnly Property SetNewWhatsThisPadding() As Boolean
            Get
                Return True
            End Get
        End Property
    End Class
End Namespace

File: Components/WinForms/CS/GridTutorials/GridDragDropInternal/GridDragDropInternal.cs
using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.XtraGrid.Demos.Tutorials {
	/// <summary>
	/// Summary description for GridDragDropInternal.
	/// </summary>
	public partial class GridDragDropInternal : TutorialControl {
		public GridDragDropInternal() {
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}
        GridHitInfo hitInfo = null;
		private void GridDragDropInternal_Load(object sender, System.EventArgs e) {
			InitData();
			gridView1.SetRowExpanded(-1, true);
		}

		private void InitData() {
			string DBFileName = DevExpress.Utils.FilesHelper.FindingFileName(Application.StartupPath,"Data\\Contacts.xml");
			if(DBFileName != "") {
				DataSet ds = new DataSet();
				ds.ReadXml(DBFileName);
				gridControl1.DataSource = ds.Tables[0];
			}
		}

		
      //<gridControl1>
		private void gridControl1_MouseDown(object sender, System.Windows.Forms.MouseEventArgs e) {
			hitInfo = gridView1.CalcHitInfo(new Point(e.X, e.Y));
			if(hitInfo.RowHandle < 0) hitInfo = null; 
		}

		private void gridControl1_MouseMove(object sender, System.Windows.Forms.MouseEventArgs e) {
			if(hitInfo == null) return;
			if(e.Button != MouseButtons.Left) return;
			Rectangle dragRect = new Rectangle(new Point(
				hitInfo.HitPoint.X - SystemInformation.DragSize.Width / 2,
				hitInfo.HitPoint.Y - SystemInformation.DragSize.Height / 2), SystemInformation.DragSize);
			if(!dragRect.Contains(new Point(e.X, e.Y))) { 
				if(hitInfo.InRow) {
					gridControl1.DoDragDrop(new object[] {hitInfo.RowHandle, gridControl1}, DragDropEffects.All);
				}
			}
		}
      //</gridControl1>

		private int GetDragObject(object sender, IDataObject data) {
			object[] obj = data.GetData(typeof(object[])) as object[];
				if(obj == null || !sender.Equals(gridControl1)) return -2; 
			if(obj[0] is Int32)
				return (int)obj[0];
			else return -1;
		}

      //<gridControl1>
		private void gridControl1_DragOver(object sender, System.Windows.Forms.DragEventArgs e) {
			GridHitInfo hi = gridView1.CalcHitInfo(gridControl1.PointToClient(new Point(e.X, e.Y)));
			int handle = GetDragObject(sender, e.Data); 
			if(hi.InRow && handle >= 0 && handle != hi.RowHandle) {
				if((e.KeyState & 8) != 0) 
					e.Effect = DragDropEffects.Copy;
				else
					e.Effect = DragDropEffects.Move;
			}
			else e.Effect = DragDropEffects.None;
		}

		private void gridControl1_DragDrop(object sender, System.Windows.Forms.DragEventArgs e) {
			GridHitInfo hi = gridView1.CalcHitInfo(gridControl1.PointToClient(new Point(e.X, e.Y)));
			int handle = GetDragObject(sender, e.Data); 
			if(handle < 0) return;
			if(hi.InRow) {
				DataRowView row = gridView1.GetRow(handle) as DataRowView;
				//the following code determines whether the target row is a group row and if a group row is detected, 
				//it retrieves the actual data row corresponding to it.
				int newHandle = hi.RowHandle;
				if(newHandle < 0) {
					gridView1.SetRowExpanded(newHandle, true, true);
					newHandle = gridView1.GetDataRowHandleByGroupRowHandle(newHandle);
				}
				DataRow rowReciever = gridView1.GetDataRow(newHandle);
				//the following code implements actual data moving or copying
                DataRowView newRow = null;
				if(e.Effect == DragDropEffects.Copy) {
					newRow = ((DataView)gridView1.DataSource).AddNew();
					for(int i = 0; i < gridView1.Columns.Count; i++) 
						newRow.Row[gridView1.Columns[i].FieldName] = row[gridView1.Columns[i].FieldName];
					row = newRow;
				}
				for(int i = 0; i < gridView1.SortInfo.GroupCount; i++) 
					row.Row[gridView1.GroupedColumns[i].FieldName] = rowReciever[gridView1.GroupedColumns[i].FieldName];
                if(newRow != null) newRow.Row["id"] = ((DataView)gridView1.DataSource).Count;
				row.EndEdit();
			}
		}
      //</gridControl1>

		private void gridView1_ShowingEditor(object sender, System.ComponentModel.CancelEventArgs e) {
			if(!MouseButtons.Equals(MouseButtons.None))
				e.Cancel = true;
		}
        public override bool SetNewWhatsThisPadding { get { return true; } }
	}
}
